sloghandler.go

```diff
@@ -0,0 +1,71 @@
+//go:build go1.21
+// +build go1.21
+
+/*
+Copyright 2023 The logr Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package logr
+
+import (
+	"context"
+	"log/slog"
+)
+
+type slogHandler struct {
+	sink        LogSink
```

Comment:
Mostly for symmetry: we are layering some high-level API on top of some low-level API. Doing the same here with `LogSink` as in `slogSink` with `slog.Handler` seemed appropriate.

> I chose Logger because it felt easier.

Is it really? Using `l.sink.Enabled/Info/Error` below is trivial and avoids calling code that doesn't add any value.

Reply:
logr.Logger is not a no-op, though.

```
logrLogger := logr.Discard()  // nil LogSink
slogLogger := logr.ToSlog()
slogLogger.Info() // -> panic (I think?)
```

Similarly, Logger.WithCallDepth() does the LogSink interface test.

Reply:
True, I need to handle a nil sink. I wasn't trying to do anything with call stack unwinding, so I didn't need that.